### DISCUSSION TITLE
ESP32 or Arduino Uno r3

### DISCUSSION THREAD

**Discussion by techsavy01** [collin80/SavvyCAN#824] [Q&A]
Hi everyone, 

Experimenting with the CAN bus, I came across Savvy CAN and had a simple question about the hardware. Can I make an ESP32 or Arudino with MCP 2515 work with Savvy CAN? if so can I get some help on which library or example sketch to use?

Any help is appreciated

**Comment by logosred** [comment]
Yes, you can use ESP32RET https://github.com/collin80/ESP32RET for the ESP32 . I'm currently writing a post on my website on how to. 

On the Uno side there aren't any libraries that bridge the gap between the Uno and SavvyCan/ can-utils over a CAN interface, there is https://github.com/autowp/arduino-canhacker but it's old and only seems to work with the nano. I have a nano on hand but haven't tried.

Good luck.

**Reply by techsavy01** [reply]
Thank you. I'm getting this error message. I'm using Esp wroom 32 and all the setups i've done my pins are set to 

MP2515 pins and ESP Wroom 32 pins
INT                                 D4
SCK                                D18
SI                                    D23
SO                                  D19
CS                                   D5
GND                               GND
VCC                                 VIN
![image](https://github.com/user-attachments/assets/dac71ede-90c3-4541-bca1-774c3b8508cf)

**Reply by logosred** [reply]
The ESP32 has a built-in CAN controller; you shouldn't be using an MCP2515 with it, really. I am pretty sure the MCP2515 board you're using is also outputting 5V logic levels, not the 3.3V that is being used by the ESP32, so you're slowly damaging it (the ESP32 being 5V tolerant is debatable).

Buy a SN65HVD230. If you can't find one, get a TJA1050 or MCP2551. Use a logic level converter (5V to 3.3V). Get the ESP32RET code; it will save you some trouble.

I'm still writing the article on how to do everything. It's the exact step-by-step guide you want. But I want to make sure it's right before I upload it to my site. You can buy the SN65HVD230 and I'll upload the guide with wiring diagrams, Arduino IDE, everything.

**Comment by techsavy01** [comment]
Sounds good I will wait until the write-up.

**Reply by logosred** [reply]
Apologize for the delay, here is the post: https://logos-red.com/blog/how-to-hack-a-car-wirelessly-sending-remote-commands

Tell me if anything isn't right or can be improved and I'll be glad to change it.

If you have any issues contact me on Discord and I'll be glad to help you out.

**Reply by NeariX67** [reply]
What pins would you use on an esp32 with I2C MCP2515?
